### PR TITLE
Feature/migration tools

### PR DESCRIPTION
**Pull request by drewtang**
## Description

Implements comprehensive migration tools for upgrading Bilan from v0.3.x to v0.4.0. Provides safe, zero-data-loss migration from vote-centric schema to flexible event-based schema with full rollback capabilities.

**Type of Change:**
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [x] 📚 Documentation update
- [ ] 🧹 Code cleanup/refactoring
- [ ] ⚡ Performance improvement
- [ ] 🔧 Build/CI changes

## Related Issues

Related to v0.4.0 implementation milestone
Part of parallel Tier 1 development effort

### Migration Tools Package
- [x] Added v0.3.x data extraction with batch processing
- [x] Added event format conversion (vote → vote_cast events)
- [x] Added v0.4.0 database schema creation with JSONB properties
- [x] Added full migration orchestration with progress tracking
- [x] Added rollback procedures with checkpoint management
- [x] Added comprehensive pre/post migration validation
- [x] Added professional CLI with 9 commands
- [x] Added TypeScript types and interfaces

### SDK Changes
- [ ] Added new APIs
- [ ] Modified existing APIs  
- [x] Updated types/interfaces
- [ ] Performance improvements
- [ ] Bug fixes

### Server Changes
- [ ] New endpoints
- [ ] Database changes
- [ ] API modifications
- [ ] Bug fixes

### Dashboard Changes
- [ ] New components
- [ ] UI improvements
- [ ] Bug fixes
- [ ] Performance improvements

### Manual Testing
- [x] Tested CLI commands with sample data
- [x] Tested migration flow end-to-end
- [x] Tested rollback procedures
- [x] Tested edge cases (empty DB, malformed data)

### Automated Testing
- [x] All existing tests pass
- [x] Added new tests for new functionality
- [x] Updated tests for modified functionality
- [x] Test coverage maintained/improved

**Test Results:**
```
✓ EventFormatConverter tests (8 passed)
✓ V3DataExtractor tests (6 passed)
✓ All migration components tested
✓ TypeScript compilation successful
```

### Migration Tools Package
- Bundle size: ~50KB (new package)
- Gzipped: ~15KB
- CLI binary: ~200KB with dependencies

### Impact
- [x] No size impact (new standalone package)
- [ ] Size decreased
- [ ] Size increased (justified below)

**Size justification:**
New standalone package - no impact on existing SDK bundle size.

## Breaking Changes

- [x] No breaking changes
- [ ] Breaking changes (documented below)

This is a new package that provides migration utilities without affecting existing APIs.

## Documentation

- [x] Updated README.md
- [x] Updated API documentation
- [x] Updated examples
- [x] Added JSDoc comments
- [x] Updated CHANGELOG.md

## Performance Impact

- [x] No performance impact
- [ ] Performance improved
- [ ] Performance decreased (justified below)

**Performance notes:**
- Batch processing for memory efficiency
- Streaming data extraction for large databases
- Optimized database queries with proper indexing
- Progress tracking with minimal overhead

### Code Quality
- [x] Code follows project style guidelines
- [x] Self-review completed
- [x] Code is well-commented
- [x] No console.log statements left in production code
- [x] TypeScript strict mode compliance

### Testing
- [x] All tests pass locally
- [x] Added tests for new functionality
- [x] Test coverage is adequate
- [x] Manual testing completed

### Documentation
- [x] Code is self-documenting
- [x] Added JSDoc for public APIs
- [x] Updated relevant documentation
- [x] Added/updated examples if needed

### Security & Privacy
- [x] No secrets or sensitive data in code
- [x] No new security vulnerabilities introduced
- [x] Privacy implications considered
- [x] Input validation added where needed

### Key Features
- **Zero data loss guarantee** with automatic checkpoints
- **Safe rollback procedures** with integrity validation
- **Comprehensive validation** (pre/post migration)
- **Professional CLI** with 9 commands
- **Batch processing** for memory efficiency
- **Progress tracking** with ETA calculations

### CLI Commands Added
```bash
bilan-migrate validate          # Validate v0.3.x database
bilan-migrate stats            # View database statistics
bilan-migrate convert          # Convert format (dry run)
bilan-migrate migrate          # Full migration
bilan-migrate validate-pre     # Pre-migration validation
bilan-migrate validate-post    # Post-migration validation
bilan-migrate rollback         # Rollback migration
bilan-migrate report           # Generate migration report
```

### Migration Flow
1. **Pre-validation** → checks readiness, creates checkpoint
2. **Data extraction** → safely reads v0.3.x data in batches
3. **Format conversion** → transforms to v0.4.0 event schema
4. **Database creation** → sets up v0.4.0 schema with indexes
5. **Data insertion** → writes converted events with validation
6. **Post-validation** → verifies integrity and performance
7. **Rollback ready** → checkpoint available for safe rollback

### Future Considerations
- Integration with v0.4.0 SDK once schema is finalized
- Performance optimizations for very large datasets
- Monitoring/observability enhancements based on user feedback

## Reviewer Checklist

- [ ] Code review completed
- [ ] Tests reviewed and adequate
- [ ] Documentation reviewed
- [ ] Breaking changes acceptable
- [ ] Performance impact acceptable
- [ ] Ready to merge
    

---

## Summary by cubic
Added a migration tools package to safely upgrade Bilan from v0.3.x to v0.4.0, converting vote-centric data to a flexible event-based schema with full validation and rollback support.

- **New Features**
  - CLI with 9 commands for validation, migration, conversion, rollback, and reporting.
  - Batch data extraction, event format conversion, and progress tracking.
  - Pre- and post-migration validation with zero data loss guarantee.
  - Rollback procedures with automatic checkpoints.

## Summary by CodeRabbit

* **New Features**
  * Introduced a migration toolkit to upgrade Bilan analytics data from version 0.3.x to 0.4.0, including CLI and programmatic interfaces.
  * Added comprehensive validation, migration, rollback, and reporting capabilities with detailed progress tracking and batch processing.
  * Provided robust pre- and post-migration checks, rollback safety, and migration reports.
  * Added a CLI tool with commands for validating, extracting, converting, migrating, rolling back, and reporting migration status.
  * Implemented event format conversion preserving metadata and ensuring data integrity.
  * Enabled dry-run mode, migration preview, and detailed migration statistics.
  * Added database managers for both source and target schemas with validation and statistics.

* **Documentation**
  * Added detailed README with installation, usage instructions, migration process overview, and safety features.

* **Tests**
  * Introduced test suites covering data extraction and event conversion logic to ensure migration accuracy and reliability.

### CODE REVIEW

File: packages/migration-tools/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    "resolveJsonModule": true,
+    "declaration": true,
+    "outDir": "./dist",
+    "rootDir": "./src",
+    "lib": ["ES2022"],
+    "useDefineForClassFields": true,
+    "isolatedModules": true,
+    "noEmitHelpers": true,
+    "importHelpers": false
+  },
```
[the remark this answers is not in the document]
Comment (drewtang) [reply packages/migration-tools/tsconfig.json]: Implemented this in commit (4740a4c)
